src/Microsoft.AspNetCore.Mvc.Core/Internal/ActionDescriptorCollectionProvider.cs

```diff
@@ -1,31 +1,53 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
-using System;
+using System.Collections.Generic;
 using System.Collections.ObjectModel;
 using System.Linq;
+using System.Threading;
 using Microsoft.AspNetCore.Mvc.Abstractions;
 using Microsoft.AspNetCore.Mvc.Infrastructure;
-using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Primitives;
 
 namespace Microsoft.AspNetCore.Mvc.Internal
 {
     /// <summary>
     /// Default implementation of <see cref="IActionDescriptorCollectionProvider"/>.
-    /// This implementation caches the results at first call, and is not responsible for updates.
     /// </summary>
     public class ActionDescriptorCollectionProvider : IActionDescriptorCollectionProvider
     {
-        private readonly IServiceProvider _serviceProvider;
+        private readonly IActionDescriptorProvider[] _actionDescriptorProviders;
+        private readonly IActionDescriptorChangeProvider[] _actionDescriptorChangeProviders;
         private ActionDescriptorCollection _collection;
+        private int _version = -1;
 
         /// <summary>
         /// Initializes a new instance of the <see cref="ActionDescriptorCollectionProvider" /> class.
         /// </summary>
-        /// <param name="serviceProvider">The application IServiceProvider.</param>
-        public ActionDescriptorCollectionProvider(IServiceProvider serviceProvider)
```

Comment:
This was intentional in the old code btw, so that we didn't have to keep the ADPs in memory after they were used. I don't think it's important. If it matters we'll find it.